### PR TITLE
chore: clean up maintenance scripts and eval outputs

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary
- add four maintenance scripts that were present locally but not tracked, including DB-safe vector maintenance helpers and a Karabiner installer for BrainBar
- make the vector maintenance scripts use repo-standard DB path resolution instead of an Etan-specific absolute path
- ignore local `tests/eval_results/` outputs and refresh the README Python test count to the verified total

## Verification
- `ruff check tests/test_lifecycle.py scripts/purge_orphaned_vectors.py scripts/rebuild_vec0_tables.py scripts/run_decay_job.py`
- `shellcheck scripts/install-karabiner-rule.sh`
- `python3 -m pytest tests/test_lifecycle.py -q` → `24 passed`
- `python3 scripts/purge_orphaned_vectors.py --help`
- `python3 scripts/rebuild_vec0_tables.py --help`
- `pytest tests/` on this branch → `5 failed, 1823 passed, 8 skipped, 6 xfailed, 2 xpassed`
- reproduced the same 5 failing tests on a clean `HEAD` archive, so the red full-suite failures predate this branch:
  - `tests/test_eval_baselines.py::TestMemoryRetrieval::test_whoop_discussion_findable`
  - `tests/test_eval_baselines.py::TestTemporalQueries::test_recent_week_chunks_exist`
  - `tests/test_eval_baselines.py::TestPromptHookEntityInjection::test_no_entity_in_generic_query`
  - `tests/test_vector_store.py::TestSchema::test_created_at_coverage`
  - `tests/test_vector_store.py::TestSearchMetadata::test_results_have_created_at`

## Notes
- deleted the untracked root session transcript `.txt` files during cleanup
- `cr review --plain` could not run because CodeRabbit returned a rate-limit error (`try after 21 minutes and 14 seconds`)

> [!NOTE]
> ### Add database maintenance scripts for purging orphaned vectors and rebuilding vec0 tables
> - Adds [`scripts/purge_orphaned_vectors.py`](https://github.com/EtanHey/brainlayer/pull/238/files#diff-bdde44726945b81f6a2490e21253c05c20c682e1da3ce3f39d9623fbae91c80a) to delete vectors in `chunk_vectors` and `chunk_vectors_binary` whose IDs no longer exist in `chunks`, operating in configurable batches with periodic WAL checkpoints.
> - Adds [`scripts/rebuild_vec0_tables.py`](https://github.com/EtanHey/brainlayer/pull/238/files#diff-0ddae23f873ca2fd045e7f9702112e06ae4d71257b25b2ba2d8f4f4422ac53bf) to reconstruct vec0 virtual tables by backing up valid embeddings, dropping and recreating the tables, and reinserting preserved data with progress reporting.
> - Adds [`scripts/run_decay_job.py`](https://github.com/EtanHey/brainlayer/pull/238/files#diff-20fda2f3e06b0ef257b5fb11e903cd719980346fe212ee574cbaabd03fcd573c) as a CLI wrapper for the decay maintenance job, supporting `--json`, `--dry-run`, and `--batch-size` flags.
> - Adds [`scripts/install-karabiner-rule.sh`](https://github.com/EtanHey/brainlayer/pull/238/files#diff-ab2662c58c91382c434f423ea722fe6c574ccb9d4e93a3fc7f16d02bb6e46d2d) to idempotently merge a BrainBar Karabiner-Elements rule into the user's config without overwriting existing rules.
> - Adds tests in [`tests/test_lifecycle.py`](https://github.com/EtanHey/brainlayer/pull/238/files#diff-d8958c006f6b1824d907540b609b4a5dca8073cc0a24ae332363d855de6ae8bc) covering the new script helpers; updates `.gitignore` to exclude `tests/eval_results/` from version control.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized a6f5cd1.</sup>
> 
>

### CODE REVIEW

**Comment by EtanHey** [comment]
@codex review

**Comment by EtanHey** [comment]
@cursor @bugbot review

**Comment by EtanHey** [comment]
@coderabbitai review

File: scripts/rebuild_vec0_tables.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+                    "SELECT embedding FROM chunk_vectors WHERE chunk_id = ?", (cid,)
+                ).fetchone()
+                if row:
+                    conn.execute(
+                        "INSERT INTO _tmp_vec_backup (chunk_id, embedding) VALUES (?, ?)",
+                        (cid, row[0]),
+                    )
+                    inserted += 1
+            except Exception as e:
+                print(f"  Warning: skip {cid}: {e}")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/rebuild_vec0_tables.py]: Addressed in `48c85e15`.

The rebuild no longer skips extraction failures. `read_embedding_or_raise()` now raises on any read error or missing row, and the rebuild aborts before dropping the vec0 table if backup extraction cannot complete cleanly. I also added a count sanity check so the backup row count must match the expected valid-vector count before the destructive step proceeds.

File: scripts/rebuild_vec0_tables.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            chunk_id TEXT PRIMARY KEY,
+            embedding BLOB
+        )
+    """)
+
+    # Read from vec0 in batches via the rowids table
+    all_ids = conn.execute(
+        "SELECT id FROM chunk_vectors_rowids WHERE id IN (SELECT id FROM chunks)"
+    ).fetchall()
+    all_ids = [r[0] for r in all_ids]
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/rebuild_vec0_tables.py]: Addressed in `48c85e15`.

I removed the `fetchall()`/materialize-all-IDs path. The rebuild now streams valid IDs via `iter_valid_chunk_ids()` and batches them with `batched(...)`, so extraction stays bounded instead of loading the full rowid table into memory up front. I also switched reinsert-from-backup to stream rows instead of `fetchall()` for the same reason.

**Comment by EtanHey** [comment]
Follow-up pushed in `48c85e15` to address Codex review:
- rebuild extraction now fails fast on missing/errored reads instead of skipping rows
- valid IDs are streamed in batches rather than materialized with `fetchall()`
- reinsert from backup now streams rows too
- added regression tests for the rebuild helpers
- refreshed README Python test count to `1,848`

Verification on the updated head:
- `ruff check tests/test_lifecycle.py scripts/rebuild_vec0_tables.py`
- `python3 -m pytest tests/test_lifecycle.py -q` → `28 passed`
- `python3 scripts/rebuild_vec0_tables.py --help`
- `pytest tests/` on updated head still shows unrelated live-db / hook failures outside this change surface; a clean `HEAD` archive reproduces the same lock/timeout class on the failing integration subset.

@codex review
@cursor @bugbot re-review

File: scripts/rebuild_vec0_tables.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            elapsed = time.time() - t0
+            rate = reinserted / elapsed if elapsed > 0 else 0
+            print(f"  Inserted {reinserted:,}/{inserted:,} ({rate:.0f}/s)", flush=True)
+
+    elapsed = time.time() - t0
+    print(f"Step 4 done: {reinserted:,} re-inserted in {elapsed:.1f}s ({errors} errors)")
+
+    # Step 5: Drop backup
+    conn.execute("DROP TABLE _tmp_vec_backup")
+    print("Backup table dropped.")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/rebuild_vec0_tables.py:163]: Addressed in `f2a8edb2`.

The rebuild now calls `ensure_restore_succeeded(...)` before dropping either backup table. If any reinsert errors occurred, the script raises and leaves the backup table intact instead of deleting the only recoverable copy.

**Comment by EtanHey** [comment]
Final hardening follow-up pushed in `f2a8edb2`:
- preserve `_tmp_vec_backup` / `_tmp_binvec_backup` when restore errors occur instead of dropping them unconditionally
- added a regression test for the restore guard

Verification on current head:
- `ruff check tests/test_lifecycle.py scripts/rebuild_vec0_tables.py`
- `python3 -m pytest tests/test_lifecycle.py -q` → `29 passed`

@codex review

**Comment by EtanHey** [comment]
Follow-up pushed in `6fff9a01` to address the remaining purge review items:
- orphan IDs are counted first and then streamed in bounded batches instead of materialized with `fetchall()`
- purge now raises on delete errors so maintenance cannot exit green after a partial cleanup
- added regression coverage for batching, streaming, and failure-on-delete-error

Verification on current head:
- `python3 -m pytest tests/test_lifecycle.py -q` -> `32 passed`
- `python3 scripts/purge_orphaned_vectors.py --help`

@codex review
@cursor @bugbot re-review

File: scripts/purge_orphaned_vectors.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    conn.execute("PRAGMA busy_timeout=10000")
+    return conn
+
+
+def get_orphan_ids(conn, vec_rowids_table: str) -> list[str]:
+    """Get all chunk IDs in the vec table that don't exist in chunks."""
+    rows = conn.execute(f"""
+        SELECT vr.id FROM {vec_rowids_table} vr
+        WHERE vr.id NOT IN (SELECT id FROM chunks)
+    """).fetchall()
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/purge_orphaned_vectors.py]: Addressed in `6fff9a01`.

The purge path no longer materializes all orphan IDs. It now does a count query for progress reporting, streams orphan IDs via `iter_orphan_ids(...)`, and batches them with `batched(...)` so memory stays bounded on large databases. I also added regression coverage in `tests/test_lifecycle.py` and re-ran `python3 -m pytest tests/test_lifecycle.py -q` (`32 passed`).

File: scripts/purge_orphaned_vectors.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+        if batch_num % CHECKPOINT_EVERY == 0:
+            conn.execute("PRAGMA wal_checkpoint(PASSIVE)")
+            print("  [WAL checkpoint]", flush=True)
+
+    elapsed = time.time() - start
+    print(f"\nDone: {deleted:,} orphans purged from {label} in {elapsed:.1f}s")
+    if errors:
+        print(f"  ({errors} errors)")
+    return deleted
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/purge_orphaned_vectors.py:122]: Addressed in `6fff9a01`.

`purge_vec_table()` now raises `RuntimeError` when any delete errors occur, so the maintenance script exits non-zero instead of reporting a partial purge as success. There is a regression test for the failure path in `tests/test_lifecycle.py`, and `python3 -m pytest tests/test_lifecycle.py -q` passed (`32 passed`).

**Comment by EtanHey** [comment]
Formatting-only follow-up pushed in `61e33ee0` to satisfy the repo lint gate after `ruff format --check` failed on the purge-script patch.

Verification on current head:
- `ruff format --check scripts/purge_orphaned_vectors.py tests/test_lifecycle.py`
- `python3 -m pytest tests/test_lifecycle.py -q` -> `32 passed`

@codex review

**Comment by EtanHey** [comment]
Follow-up pushed in `57b3d801` for the new Macroscope review round:
- purge now re-queries one bounded orphan batch at a time instead of iterating a live cursor across rows being deleted
- the loop stops after a failed delete so the same orphan cannot be reselected forever
- `main()` now closes the SQLite connection in a `finally` block even when purge raises

Verification on current head:
- `ruff check scripts/purge_orphaned_vectors.py tests/test_lifecycle.py`
- `ruff format --check scripts/purge_orphaned_vectors.py tests/test_lifecycle.py`
- `python3 -m pytest tests/test_lifecycle.py -q` -> `32 passed`
- `python3 scripts/purge_orphaned_vectors.py --help`

@codex review

File: scripts/purge_orphaned_vectors.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    if total == 0:
+        print("Nothing to purge.")
+        return 0
+
+    deleted = 0
+    errors = 0
+    batch_num = 0
+    start = time.time()
+
+    for batch in batched(iter_orphan_ids(conn, vec_rowids_table), BATCH_SIZE):
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/purge_orphaned_vectors.py]: Addressed in `57b3d801`.

I replaced the live-cursor batching with bounded re-queries via `get_orphan_batch(...)`, so each loop iteration fetches at most one fresh batch of orphan IDs before deleting them. That keeps memory bounded without mutating the same cursor result set being traversed.

File: scripts/purge_orphaned_vectors.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+    elapsed = time.time() - start
+    print(f"\nDone: {deleted:,} orphans purged from {label} in {elapsed:.1f}s")
+    if errors:
+        print(f"  ({errors} errors)")
+        raise RuntimeError(f"Failed to purge {errors} orphaned vectors from {label}")
+    return deleted
+
+
+def main() -> None:
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/purge_orphaned_vectors.py:125]: Addressed in `57b3d801`.

`main()` now wraps the whole purge flow in `try/finally`, so the SQLite connection is always closed even if `purge_vec_table()` raises. I added regression coverage for that close-on-error path in `tests/test_lifecycle.py`.

**Comment by EtanHey** [comment]
Follow-up pushed in `a6f5cd1c` for the latest Codex review item:
- orphan-count and orphan-batch queries now use `NOT EXISTS` instead of NULL-sensitive `NOT IN`
- added a regression test that asserts the purge queries stay NULL-safe

Verification on current head:
- `ruff check scripts/purge_orphaned_vectors.py tests/test_lifecycle.py`
- `ruff format --check scripts/purge_orphaned_vectors.py tests/test_lifecycle.py`
- `python3 -m pytest tests/test_lifecycle.py -q` -> `33 passed`
- `python3 scripts/purge_orphaned_vectors.py --help`

@codex review

File: scripts/purge_orphaned_vectors.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    ).fetchone()
+    return int(row[0]) if row else 0
+
+
+def iter_orphan_ids(conn, vec_rowids_table: str) -> Iterator[str]:
+    """Stream chunk IDs in the vec table that don't exist in chunks."""
+    for row in conn.execute(
+        f"""
+        SELECT vr.id FROM {vec_rowids_table} vr
+        WHERE vr.id NOT IN (SELECT id FROM chunks)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/purge_orphaned_vectors.py]: Addressed in `a6f5cd1c`.

Both orphan queries now use `NOT EXISTS` instead of `NOT IN`, so a stray `NULL` in `chunks.id` cannot cause silent under-deletion. I also added a regression test that asserts the purge queries stay NULL-safe, and the focused lifecycle slice passed (`33 passed`).
